staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_annotations.go

```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/audit/policy"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithAuditAnnotations decorates a http.Handler with a map that is merged with
+// the audit.Event.Annotations map.  This allows layers that run before WithAudit
+// (such as authentication) to assert annotations.
+// If sink or audit policy is nil, no decoration takes place.
+func WithAuditAnnotations(handler http.Handler, sink audit.Sink, policy policy.Checker) http.Handler {
+	// no need to wrap if auditing is disabled
+	if sink == nil || policy == nil {
```

Comment:
Fold into WithAudit so other external apiservers that have audit enabled get this for free? I see no reason why we don't always want this.

Reply:
I do not understand the request.

`WithAuditAnnotations` is a distinct wrapper because we need something early in the chain before `WithAudit`.  `WithAudit` cannot be moved earlier in the chain because it relies on `WithAuthentication`.  This basically breaks the "import cycle" that authentication and audit have with each other.

Reply:
Ah, ordering...

Reply:
This is vaguely reminiscent of https://github.com/kubernetes/kubernetes/issues/86063

Reply:
Yeah, we may have taken this wrapping of handlers and context too far...